#!/bin/sh /usr/share/dpatch/dpatch-run 

## opt-44_rotor.dpatch by Thomas Bergwinkl <Bergwinkl.Thomas@vr-web.de>
##
## extracted from the rotor plugin 0.1.1
## original filename: vdr-1.3.31-Rotor.diff
## 
## All lines beginning with `## DP:' are a description of the patch. 
## DP: This patch is needed for the rotor plugin.

@DPATCH@
diff -ubw vdr-1.3.31/device.h vdr-1.3.31-rotor/device.h
--- vdr-1.3.31/device.h	2005-08-21 10:52:20.000000000 +0200
+++ vdr-1.3.31-rotor/device.h	2005-09-05 12:23:14.000000000 +0200
@@ -21,6 +21,7 @@
 #include "spu.h"
 #include "thread.h"
 #include "tools.h"
+#include <linux/dvb/frontend.h>
 
 #define MAXDEVICES         16 // the maximum number of devices in the system
 #define MAXPIDHANDLES      64 // the maximum number of different PIDs per device
@@ -240,6 +241,7 @@
   virtual bool HasProgramme(void);
          ///< Returns true if the device is currently showing any programme to
          ///< the user, either through replaying or live.
+  virtual bool SendDiseqcCmd(dvb_diseqc_master_cmd cmd) {return false;}
 
 // PID handle facilities
 
diff -ubw vdr-1.3.31/dvbdevice.c vdr-1.3.31-rotor/dvbdevice.c
--- vdr-1.3.31/dvbdevice.c	2005-08-21 11:17:20.000000000 +0200
+++ vdr-1.3.31-rotor/dvbdevice.c	2005-09-05 12:25:01.000000000 +0200
@@ -70,6 +70,7 @@
 class cDvbTuner : public cThread {
 private:
   enum eTunerStatus { tsIdle, tsSet, tsTuned, tsLocked, tsCam };
+  bool SendDiseqc;
   int fd_frontend;
   int cardIndex;
   fe_type_t frontendType;
@@ -82,6 +83,7 @@
   cMutex mutex;
   cCondVar locked;
   cCondVar newSet;
+  dvb_diseqc_master_cmd diseqc_cmd;
   bool GetFrontendEvent(dvb_frontend_event &Event, int TimeoutMs = 0);
   bool SetFrontend(void);
   virtual void Action(void);
@@ -90,12 +92,14 @@
   virtual ~cDvbTuner();
   bool IsTunedTo(const cChannel *Channel) const;
   void Set(const cChannel *Channel, bool Tune, bool UseCa);
+  bool SendDiseqcCmd(dvb_diseqc_master_cmd cmd);
   bool Locked(int TimeoutMs = 0);
   };
 
 cDvbTuner::cDvbTuner(int Fd_Frontend, int CardIndex, fe_type_t FrontendType, cCiHandler *CiHandler)
 {
   fd_frontend = Fd_Frontend;
+  SendDiseqc=false;
   cardIndex = CardIndex;
   frontendType = FrontendType;
   ciHandler = CiHandler;
@@ -148,6 +152,15 @@
   return tunerStatus >= tsLocked;
 }
 
+bool cDvbTuner::SendDiseqcCmd(dvb_diseqc_master_cmd cmd)
+{
+  if (frontendType!=FE_QPSK || SendDiseqc)
+    return false;
+  diseqc_cmd=cmd;
+  SendDiseqc=true;
+  return true;
+}
+
 bool cDvbTuner::GetFrontendEvent(dvb_frontend_event &Event, int TimeoutMs)
 {
   if (TimeoutMs) {
@@ -299,6 +312,10 @@
         bool hasEvent = GetFrontendEvent(event, 1);
 
         cMutexLock MutexLock(&mutex);
+        if (SendDiseqc) {
+           CHECK(ioctl(fd_frontend, FE_DISEQC_SEND_MASTER_CMD, &diseqc_cmd));
+           SendDiseqc=false;
+           }
         switch (tunerStatus) {
           case tsIdle:
                break;
@@ -863,6 +880,11 @@
   return dvbTuner ? dvbTuner->Locked(TimeoutMs) : false;
 }
 
+bool cDvbDevice::SendDiseqcCmd(dvb_diseqc_master_cmd cmd)
+{
+  return dvbTuner->SendDiseqcCmd(cmd);
+}
+
 int cDvbDevice::GetAudioChannelDevice(void)
 {
   if (HasDecoder()) {
diff -ubw vdr-1.3.31/dvbdevice.h vdr-1.3.31-rotor/dvbdevice.h
--- vdr-1.3.31/dvbdevice.h	2005-08-20 17:20:15.000000000 +0200
+++ vdr-1.3.31-rotor/dvbdevice.h	2005-09-05 12:23:14.000000000 +0200
@@ -66,6 +66,7 @@
   virtual bool SetChannelDevice(const cChannel *Channel, bool LiveView);
 public:
   virtual bool HasLock(int TimeoutMs = 0);
+  virtual bool SendDiseqcCmd(dvb_diseqc_master_cmd cmd);
 
 // PID handle facilities
 
